PXB-2506 Error out if backup is not prepared during copy-back

rahulmalik87 opened the pull request:
https://jira.percona.com/browse/PXB-2506

Copy back allowed restoring backup which are not fully prepared.
for example prepared with *apply-log-only.*

This could end in restoring wrong backup.

Fix
add check to validate if backup is fully prepared.

altmannmarcelo reviewed (changes requested):
Don't we have a place at the beginning of --prepare that tells if the target has been prepared already? We could use the same function logic and invoke it at [copy|move]-back.

Can you check that?

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_copy.cc:
```diff
@@ -2194,10 +2195,23 @@ bool copy_back(int argc, char **argv) {
       !directory_exists(innobase_data_home_dir, true)) {
     return (false);
   }
+  if (!xtrabackup_read_metadata(metadata_path)) {
+    msg("xtrabackup: Error: failed to read metadata from '%s'\n",
+        metadata_path);
+    exit(EXIT_FAILURE);
+  }
```
I think exiting is not right. Maybe throw a warning and proceed with the old behavior.

rahulmalik87 replied:
It is added because in 8.0 we can't start the server which is not fully prepared.

satya-bodapati replied:
Are you sure? Can you try a full backup and try restoring without prepare?

rahulmalik87 replied:
I tried but I can't reproduce the case which caused copy back to fail if it is just prepare without apply-log. So what to do you suggest just add warning?

satya-bodapati replied:
Yeah, at least a warning for a few releases before making it throw an error.

rahulmalik87 replied:
]for latest release it crashes even on empty directory. 

 Assertion failure: buf0flu.cc:1060:skip_lsn_check || mach_read_from_8(page + 16) <= newest_lsn thread 140061461489408
2022-03-03T09:56:39.786364Z 0 [Note] [MY-010309] [

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -253,9 +253,6 @@ ulint xtrabackup_log_copy_interval = 1000;
 char *log_ignored_opt = NULL;
 
 /* === metadata of backup === */
-#define XTRABACKUP_METADATA_FILENAME "xtrabackup_checkpoints"
```
why this #define is removed?

rahulmalik87 replied:
moved to xtrabackup.h

rahulmalik87 replied:
changed the approach

rahulmalik87 replied:
> > I am using the same method used by prepare xtrabackup_read_metadata, just need to access inside backup_copy.cc
> 
> Why not do it on xtrabackup.cc as soon as we read --copy/move-back parameter? For example at https://github.com/percona/percona-xtrabackup/blob/percona-xtrabackup-8.0.26-18/storage/innobase/xtrabackup/src/xtrabackup.cc#L7790

Thanks for suggestion. Implemented the changes only in xtrabackup.cc

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -253,9 +253,6 @@ ulint xtrabackup_log_copy_interval = 1000;
 
 /* === metadata of backup === */
-#define XTRABACKUP_METADATA_FILENAME "xtrabackup_checkpoints"
-static char metadata_type_str[30] = ""; /*[full-backuped|log-applied|
-                                        full-prepared|incremental]*/
```
just removing static is fine? Any reason for moving it down?

rahulmalik87 replied:
moved to xtrabackup.h

rahulmalik87 commented:
> Don't we have a place at the beginning of --prepare that tells if the target has been prepared already? We could use the same function logic and invoke it at [copy|move]-back.
> 
> Can you check that?

I am using the same method used by prepare xtrabackup_read_metadata, just need to access inside backup_copy.cc

altmannmarcelo reviewed (changes requested):
> I am using the same method used by prepare xtrabackup_read_metadata, just need to access inside backup_copy.cc

Why not do it on xtrabackup.cc as soon as we read --copy/move-back parameter?
For example at https://github.com/percona/percona-xtrabackup/blob/percona-xtrabackup-8.0.26-18/storage/innobase/xtrabackup/src/xtrabackup.cc#L7790

Commit:
PXB-2506 Error out if backup is not prepared during copy-back
https://jira.percona.com/browse/PXB-2506
Copy back allowed restoring backup which are not fully prepared.
for example prepared with *apply-log-only.*
This could end in restoring wrong backup.
Fix
add check to validate if backup is fully prepared.

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -7965,6 +7961,16 @@ int main(int argc, char **argv) {
+
+    /* warning if target is not prepared */
+    read_metadata();
+    std::cout << metadata_type_str << " metadata_type_str " << std::endl;
```
This is not at all the correct way after the new PXB logging mechanism ;)

rahulmalik87 replied:
removed :)

altmannmarcelo reviewed (approved):
Apart from the logging issue Satya pointed out LGTM.
 
Have you run local tests?
Apart from that, lets pay some attention next week when full jenkins run on trunk to see if there are no edge cases related to specific platforms.

altmannmarcelo commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -7965,6 +7961,15 @@ int main(int argc, char **argv) {
       exit(EXIT_FAILURE);
     }
+
+    /* warning if target is not prepared */
```
Minor issue.
Can you adjust the comment? We are aborting the operation, not warning.

rahulmalik87 replied:
fixed

satya-bodapati commented:
@rahulmalik87 please submit on Jenkins

rahulmalik87 commented:
> @rahulmalik87 please submit on Jenkins. 

I executed the test on local machine. Below is the result. 

Spent 21304 of 5371 seconds executing testcases
SUMMARY: 332 run, 273 successful, 54 skipped, 5 failed

Skipped tests:  binlog.bug1250375 keyring.innodb_encryption_mix_plugins keyring.innodb_keyring_kmip_component keyring.innodb_keyring_vault_auto_v1 keyring.innodb_keyring_vault_auto_v2 keyring.innodb_keyring_vault_v1 keyring.innodb_keyring_vault_v2 keyring.keyring_encryption keyring.pxb-1793 keyring.shared_tablespace_encryption rocksdb.compress_encrypt rocksdb.compress rocksdb.cross_engine_trx rocksdb.datadir rocksdb.ddl rocksdb.local_max_age rocksdb.local rocksdb.log_status rocksdb.stream_compress_encrypt rocksdb.stream_compress rocksdb.stream rocksdb.wal_dir_abs_archive rocksdb.wal_dir_abs rocksdb.wal_dir_rel main.backup_locks main.bug1192834-purge main.bug1192834-rollback main.bug1227240 main.bug1347698 main.bug1394493 main.bug1415191 main.bug1418584 main.bug1475487 main.bug1533722 main.bug382742 main.bug759225 main.bug976945 main.bug983720_galerainfo main.bug983720_lrudump main.distribution_message main.ib_doublewrite main.ib_lru_dump_basic main.ib_lru_dump_rsync main.ib_lru_dump_stream main.ib_part_tf_myisam main.innodb_fast_checksum main.innodb_log_checksum_algorithm main.pxb-1901 main.pxb-1824 main.pxb-1914 main.xb_apply_archived_logs main.xbcloud main.xb_galera_info main.xb_galera_sst

Failed tests:  keyring.innodb_redo_log_encrypt main.PXB-1548_basic main.PXB-1548_rsync_basic main.pxb-1870 main.xb_compress_lz4

The failed test are because of lz4 not installed on system and dump of innodb buffer pool which are not related to changes. 
Is it sufficient?
